### PR TITLE
[PHEE-574] Bulk flaky TCs fix and Use Await instead of thread.sleep

### PR DESCRIPTION
**Pull request by dhruvsonagara**
## Description

This PR contains below tickets changes also 
* [PHEE-544 Replace Thread.sleep()/wait() in Int Tests with event-driven functi…](https://fynarfin.atlassian.net/browse/PHEE-544?atlOrigin=eyJpIjoiYzg3MjI1OTI0ZDIxNGQ2YWJlNWU3Yzk1OWVjZDBhODkiLCJwIjoiamlyYS1zbGFjay1pbnQifQ)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!
- [x] Followed the PR title naming convention mentioned above.

- [x] Acknowledge that we will not merge PRs that are not passing the checks ("green") - it is your (author's) responsibility to get a proposed PR to pass all the checks, not primarily the project's maintainers.

- [x] The PR title should include a JIRA ticket

- [x] Design-related bullet points or design document links related to this PR added in the description above.

- [x] Updated corresponding Postman Collection or API documentation for the changes in this PR.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Add required Swagger annotation and update API documentation with details of any API changes if applicable

- [ ] Followed the naming conventions as given in https://docs.google.com/document/d/1Q4vaMSzrTxxh9TS0RILuNkSkYCxotuYk1Xe0CMIkkCU/edit?usp=sharing

- [ ] Followed coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

FYI our guidelines for code reviews same as https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/BatchApiStepDef.java [hunk 1]
```diff
@@ -521,65 +545,80 @@ public void iAmAbleToParseSubBatchSummaryResponse() {
...
     public void matchTotalSubBatchTxnAndBatchTxnCount() {
-        assertThat(scenarioScopeState.batchAndSubBatchSummaryResponse).isNotNull();
-        assertThat(Integer.parseInt(scenarioScopeState.batchAndSubBatchSummaryResponse.getTotalSubBatches())).isGreaterThan(1);
-        long batchTotal = scenarioScopeState.batchAndSubBatchSummaryResponse.getTotal();
-        long subBatchTotal = 0L;
-        for (SubBatchSummary subBatchSummary : scenarioScopeState.batchAndSubBatchSummaryResponse.getSubBatchSummaryList()) {
-            subBatchTotal += subBatchSummary.getTotal();
-        }
-        assertThat(batchTotal).isEqualTo(subBatchTotal);
+        await().atMost(awaitMost, SECONDS).pollInterval(pollInterval, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/BatchApiStepDef.java:586]: we dont need await here.

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/BatchAuthorizationStepDef.java [hunk 2]
```diff
@@ -49,7 +51,9 @@ public void iShouldGetBatchStatusAs(String status) {
 
     @Then("I should be able to verify that the {string} method to {string} endpoint received a request with authorization status")
     public void iShouldBeAbleToVerifyThatTheMethodToEndpointReceivedARequestWithAuthorizationStatus(String httpMethod, String endpoint) {
-        verify(postRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.status", equalTo("Y"))));
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/BatchAuthorizationStepDef.java:54]: use poll intervel and also add an assertion

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/ChannelCollectionStepDef.java [hunk 3]
```diff
@@ -62,13 +64,13 @@ public void iCallTheTxnAPIWithTransactionId() throws InterruptedException {
         }
         requestSpec.queryParam("transactionId", scenarioScopeState.transactionId);
 
-        Thread.sleep(10000);
+        await().atMost(awaitMost, SECONDS).pollInterval(pollInterval, SECONDS).untilAsserted(() -> {
+            scenarioScopeState.response = RestAssured.given(requestSpec).baseUri(operationsAppConfig.operationAppContactPoint).expect()
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/ChannelCollectionStepDef.java:68]: missing assert statement inside

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/GSMATransferStepDef.java [hunk 4]
```diff
@@ -495,49 +497,54 @@ public void iCallTheRegisterBeneficiaryAPIWithExpectedStatusOfAndCallbackStub(in
...
-        scenarioScopeState.response = RestAssured.given(requestSpec).header("Content-Type", "application/json")
-                .header("X-Registering-Institution-ID", registeringInstitutionId)
-                .header("X-CallbackURL", identityMapperConfig.callbackURL + stub).queryParam("payeeIdentity", payeeIdentity)
-                .queryParam("paymentModality", "01").queryParam("requestId", requestId)
-                .baseUri(identityMapperConfig.identityMapperContactPoint).expect()
-                .spec(new ResponseSpecBuilder().expectStatusCode(expectedStatus).build()).when()
-                .get(identityMapperConfig.accountLookupEndpoint).andReturn().asString();
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
+            requestId = generateUniqueNumber(10);
+            RequestSpecification requestSpec = Utils.getDefaultSpec();
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/GSMATransferStepDef.java:502]: Add an assert statement inside

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/IdentityMapperStepDef.java [hunk 5]
```diff
@@ -163,24 +165,24 @@ public void iCallTheAccountLookupAPITimesWithExpectedStatusOf(int count, int exp
                     .spec(new ResponseSpecBuilder().expectStatusCode(expectedStatus).build()).when()
                     .get(identityMapperConfig.accountLookupEndpoint).andReturn().asString();
 
-            try {
-                Thread.sleep(2000);
-            } catch (InterruptedException e) {
-                throw new RuntimeException(e);
-            }
+            await().atMost(awaitMost, SECONDS).until(() -> true);
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/IdentityMapperStepDef.java:168]: please check this, we do not need this line here

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/IdentityMapperStepDef.java [hunk 6]
```diff
@@ -163,24 +165,24 @@ public void iCallTheAccountLookupAPITimesWithExpectedStatusOf(int count, int exp
...
 
         }
     }
 
     @Then("I should be able to verify that the {string} method to {string} endpoint received a request with required parameter in body")
     public void iShouldBeAbleToVerifyThatTheMethodToEndpointReceivedRequestWithASpecificBody(String httpmethod, String endpoint) {
-        verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.registerRequestID", equalTo(requestId))));
-        verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.numberFailedCases", equalTo("0"))));
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/IdentityMapperStepDef.java:175]: add a valid assert here

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/IdentityMapperStepDef.java [hunk 7]
```diff
@@ -163,24 +165,24 @@ public void iCallTheAccountLookupAPITimesWithExpectedStatusOf(int count, int exp
...
+            verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.registerRequestID", equalTo(requestId))));
+            verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.numberFailedCases", equalTo("0"))));
+        });
     }
 
     @Then("I should be able to verify that the {string} method to {string} endpoint received a request with same payeeIdentity")
     public void iShouldBeAbleToVerifyThatTheMethodToEndpointReceivedARequestWithSamePayeeIdentity(String httpmethod, String endpoint) {
-        verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.payeeIdentity", equalTo(payeeIdentity))));
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/IdentityMapperStepDef.java:183]: check this for assert

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/MockServerStepDef.java [hunk 8]
```diff
@@ -94,20 +96,22 @@ public void hitStubEndpoint(HttpMethod httpMethod, String endpoint, int expected
 
     @Then("I should be able to verify that the {httpMethod} method to {string} endpoint received {int} request")
     public void verifyStub(HttpMethod httpMethod, String endpoint, int numberOfRequest) {
-        switch (httpMethod) {
-            case GET -> {
-                verify(numberOfRequest, getRequestedFor(urlEqualTo(endpoint)));
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/MockServerStepDef.java:99]: check if the verify can be the asser here, also add intervel

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/OperationsStepDef.java [hunk 9]
```diff
@@ -52,8 +54,10 @@ public void operationsBatchTestTearDown() {
 
     @When("I call the batches endpoint with expected status of {int}")
     public void simpleBatchesApiCallWithNoHeader(int expectedStatus) {
-        log.info("Query params: {}", scenarioScopeState.batchesEndpointQueryParam);
-        callBatchesEndpoint(expectedStatus, scenarioScopeState.batchesEndpointQueryParam);
+        await().atMost(awaitMost, SECONDS).pollDelay(pollDelay, SECONDS).pollInterval(pollInterval, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/OperationsStepDef.java:57]: we dont need await here

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/OperationsStepDef.java [hunk 10]
```diff
@@ -65,8 +69,10 @@ public void assertCountOfBatches(int expectedCount) {
 
     @Then("I am able to parse batch paginated response into DTO")
     public void parseBatchPaginatedDto() {
-        assertThat(scenarioScopeState.response).isNotNull();
-        parseBatchesResponse(scenarioScopeState.response);
+        await().atMost(awaitMost, SECONDS).pollInterval(pollInterval, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/OperationsStepDef.java:72]: we dont need await here

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/PayerFundTransferStepDef.java [hunk 11]
```diff
@@ -415,13 +418,16 @@ public void iCallTheTransferAPIWithTransactionId() throws InterruptedException {
 
     @Then("I check for error related to {}")
     public void checkForError(String action) {
-        JsonObject jsonObject = JsonParser.parseString(scenarioScopeState.response).getAsJsonObject();
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/PayerFundTransferStepDef.java:421]: no need to add await here

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 12]
```diff
@@ -85,15 +87,17 @@ public void iCreateAnIdentityMapperDTOForRegisterBeneficiary() {
...
-        scenarioScopeState.response = RestAssured.given(requestSpec).header("Content-Type", "application/json")
-                .header("X-CallbackURL", identityMapperConfig.callbackURL + stub)
-                .header("X-Registering-Institution-ID", registeringInstitutionId)
-                .baseUri(voucherManagementConfig.voucherManagementContactPoint).body(createVoucherBody).expect()
-                .spec(new ResponseSpecBuilder().expectStatusCode(expectedStatus).build()).when()
-                .post(voucherManagementConfig.createVoucherEndpoint).andReturn().asString();
-
-        logger.info("Voucher Response: {}", scenarioScopeState.response);
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
+            RequestSpecification requestSpec = Utils.getDefaultSpec();
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:91]: add an assert inside this method, add poll intervel

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 13]
```diff
@@ -264,16 +271,19 @@ public void iCallTheRedeemVoucherAPIWithExpectedStatusOf(int responseCode) {
 
     @Then("I can assert that redemption was successful by asserting the status in response")
     public void iCanAssertThatRedemptionWasSuccessfulByAssertingTheStatusInResponse() {
-        try {
-            JsonNode rootNode = objectMapper.readTree(redeemVoucherResponseBody);
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:274]: add assert inside

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 14]
```diff
@@ -295,12 +305,10 @@ public void createAndActivateVoucher() {
...
         mockServerStepDef.startStub("/activateVoucher", PUT, 200);
         iCallTheVoucherCreateAPIWithExpectedStatusOf(202, "/createVoucher");
-        try {
-            Thread.sleep(5000);
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e);
-        }
-        iShouldBeAbleToExtractResponseBodyFromCallback();
+        await().atMost(awaitMost, SECONDS).pollInterval(pollInterval, SECONDS).until(() -> {
+            iShouldBeAbleToExtractResponseBodyFromCallback();
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:309]: add assert statement

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 15]
```diff
@@ -313,11 +321,7 @@ public void createVoucher() {
         mockServerStepDef.startStub("/createVoucher", PUT, 200);
         mockServerStepDef.startStub("/activateVoucher", PUT, 200);
         iCallTheVoucherCreateAPIWithExpectedStatusOf(202, "/createVoucher");
-        try {
-            Thread.sleep(5000);
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e);
-        }
+        await().atMost(awaitMost, SECONDS).pollInterval(pollInterval, SECONDS).until(() -> true);
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:324]: we dont need this sleep here, add it inside iShouldBeAbleToExtractResponseBodyFromCallback() instead

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 16]
```diff
@@ -358,16 +362,18 @@ public void iCanCreateAnVoucherRequestDTOForVoucherSuspension() {
...
-                .queryParam("command", "suspend").header("X-Registering-Institution-ID", registeringInstitutionId)
-                .header("X-CallbackURL", identityMapperConfig.callbackURL + stub).header("X-Program-ID", "")
-                .baseUri(voucherManagementConfig.voucherManagementContactPoint).body(suspendVoucherBody).expect()
-                .spec(new ResponseSpecBuilder().expectStatusCode(responseCode).build()).when()
-                .put(voucherManagementConfig.voucherLifecycleEndpoint).andReturn().asString();
-
-        redeemVoucherResponseBody = scenarioScopeState.response;
-        logger.info("Voucher Response: {}", scenarioScopeState.response);
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
+            RequestSpecification requestSpec = Utils.getDefaultSpec();
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:366]: add poll interval and add assert statement

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 17]
```diff
@@ -446,21 +452,25 @@ public void iCanAssertThatRedemptionWasUnsuccessfulByAssertingTheStatusInRespons
 
     @Then("I should be able to assert response body from callback on {string}")
     public void iShouldBeAbleToAssertResponseBodyFromCallback(String endpoint) {
-        verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.registerRequestId", equalTo(requestId))));
-        verify(putRequestedFor(urlEqualTo(endpoint)).withRequestBody(matchingJsonPath("$.numberFailedCases", equalTo("0"))));
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:455]: check this block for assert

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java [hunk 18]
```diff
@@ -446,21 +452,25 @@ public void iCanAssertThatRedemptionWasUnsuccessfulByAssertingTheStatusInRespons
...
-        RequestSpecification requestSpec = Utils.getDefaultSpec();
-        scenarioScopeState.response = RestAssured.given(requestSpec).header("Content-Type", "application/json")
-                .header("X-Registering-Institution-ID", registeringInstitutionId)
-                .baseUri(voucherManagementConfig.voucherManagementContactPoint).expect()
-                .spec(new ResponseSpecBuilder().expectStatusCode(responseCode).build()).when()
-                .get(voucherManagementConfig.fetchVoucherEndpoint + "/" + serialNumber).andReturn().asString();
-
-        fetchVoucherResponseBody = scenarioScopeState.response;
-        logger.info("Voucher Response: {}", scenarioScopeState.response);
+        await().atMost(awaitMost, SECONDS).untilAsserted(() -> {
```
Comment (fynmanoj) [inline src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:463]: here await may not be needed. Please check

**Review by fynmanoj — CHANGES_REQUESTED** [review]
changes requested

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/BillPayStepDef.java [hunk 19]
```diff
@@ -401,84 +413,91 @@ public void iShouldGetDataInResponse() throws JSONException {
...
     }
 
     @Then("I should be able to extract response body from callback for bill notification with missing values")
     public void iShouldBeAbleToExtractResponseBodyFromCallbackForBillNotificationWithMissingValues() throws JSONException {
-        JSONObject jsonObject = new JSONObject(scenarioScopeState.response);
-        scenarioScopeState.transactionId = jsonObject.getString("transactionId");
-        assertThat(
-                scenarioScopeState.transactionId.equals("Invalid Request: Mandatory Fields Missing, Missing field is billInquiryRequestId"))
-                .isTrue();
+        await().atMost(awaitMost, SECONDS).pollDelay(pollDelay, SECONDS).pollInterval(pollInterval, SECONDS).untilAsserted(() -> {
```
[the remark this answers is not in the document]
Comment (dhruvsonagara) [reply src/test/java/org/mifos/integrationtest/cucumber/stepdef/BillPayStepDef.java:455]: BP-003 has sleep for this test exactly before this step

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java (hunk 18 above)
Comment (dhruvsonagara) [reply src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:463]: Fetch voucher API test requires 3 second of wait.

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/BatchApiStepDef.java (hunk 1 above)
Comment (dhruvsonagara) [reply src/test/java/org/mifos/integrationtest/cucumber/stepdef/BatchApiStepDef.java:586]: Done

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/ErrorCodeStepDef.java [hunk 20]
```diff
@@ -113,7 +116,7 @@ public void iShouldPollTheTransferQueryEndpointWithTransactionIdUntilErrorInform
     }
 
     public void iWillSleepForSecs(int time) throws InterruptedException {
-        Thread.sleep(time);
+        await().atMost(time, MILLISECONDS).pollInterval(pollInterval, SECONDS).until(() -> true);
```
[the remark this answers is not in the document]
Comment (dhruvsonagara) [reply src/test/java/org/mifos/integrationtest/cucumber/stepdef/ErrorCodeStepDef.java:119]: This method is required as it is used 3 times in the the same stepdef

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java (hunk 13 above)
Comment (dhruvsonagara) [reply src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:274]: Already assertion is present in try catch block

File: src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java (hunk 14 above)
Comment (dhruvsonagara) [reply src/test/java/org/mifos/integrationtest/cucumber/stepdef/VoucherManagementStepDef.java:309]: Here it calls the extract response from callback and assert is present there.

**Pull request merged**